Assets/Scripts/Road/ObstacleSpawner.cs

```diff
@@ -0,0 +1,88 @@
+using System.Collections.Generic;
+using UnityEngine;
+using Random = UnityEngine.Random;
+
+/// <summary>
+/// Spawns random gameObjects at random x-position
+/// </summary>
+public class ObstacleSpawner : MonoBehaviour {
+    [Header("Trees")]
+    [SerializeField]
+    private int minToSpawn = 3;
+    [SerializeField]
+    private int maxToSpawn = 10;
+    [SerializeField]
+    [Tooltip("It's how close a neighbour-tree can be placed, it should be the same as player movement")]
+    private int gameObjectPlacedPerUnit = 3;
+    [SerializeField]
+    private List<GameObject> prefabsToSpawn = new List<GameObject>();
+
+    // Data information
+    private int _maxPositionZ;
+    private Transform _parentTransform;
+    private List<Vector3> _placedGameObjectPositions;
+
+    /// <summary>
+    /// Returns random GameObject from list
+    /// </summary>
+    /// <returns></returns>
+    private GameObject GetRandomObject() {
+        return prefabsToSpawn[Random.Range(0, prefabsToSpawn.Count)];
+    }
+
+    /// <summary>
+    /// Gets random-position that is divided by 3, and is not already in list
+    /// </summary>
+    /// <returns></returns>
+    private int GetRandomXPosition() {
+        // Can spawn in x-position from -30 to 30, and removes gameObjectPlacedPerUnit on both sides for some padding
+        // Else half the trees at the edges spawns outside of the map  
+        int randomXPosition = Random.Range(-_maxPositionZ + gameObjectPlacedPerUnit, _maxPositionZ - gameObjectPlacedPerUnit + 1);
+        bool isPositionInList = _placedGameObjectPositions
+            .Exists(position => position.z.Equals(randomXPosition));
+
+        if (randomXPosition % gameObjectPlacedPerUnit == 0 && !isPositionInList) {
+            return randomXPosition;
+        }
+
+        return GetRandomXPosition();
```

Comment:
woahhhhhhhhhhhh smart

Reply:
hell yeaaah recursive methods is the way to go